### DISCUSSION TITLE
Matrix profiles and non-stationary time series forecasting

### DISCUSSION THREAD

**Discussion by anovv** [TDAmeritrade/stumpy#472] [Help: Best Practices]
I have recently discovered all the work around matrix profiles and first of all wanted to thank authors for this project.

The question I have is regarding the way matrix profiles fit in forecasting techniques of non-stationary time series. To my knowledge, the conventional way of forecasting (pattern search) time series of non-stationary data requires deriving stationary data first (e.g. using differencing) and then, using hand-crafted features (usually variation of rolling window aggregates such as Exponential Moving Average, etc.) and labeling output data, feed them into some form supervised learning framework. What is not clear to me is where exactly matrix profiles fit in this pipeline. Given this context, the questions are:

- Do we still need to remove non-stationarity from the data, or using matrix profiles derived features on raw non-stationary dataset as inputs for ML models will somehow bypass the problem of non-stationarity?
- Do matrix profile derived features make rolling window features (such as EMA) redundant, as my understanding is such that Matrix profiles are essentially a superset here and carry more information?
-  What are the examples of using Matrix profiles in conjunction with supervised learning models (e.g. Neural Nets)?
-  Can Matrix profiles be used as a standalone forecasting framework, without ML models. If so, how?

In general, any links to relevant research covering this topics would be greatly appreciated. Thank you!

**Comment by JaKasb** [comment]
The Matrix Profile is a tool for pattern-recognition and motif-search in timeseries.
You could use the knowledge of pattern occurence and timestamps for a forecasting algorithm.
There are some papers on Google scholar about this topic.
https://scholar.google.com/scholar?&q=matrix-profile+forecast+forecasting

Matrix Profile uses the Pearson Correlation Coefficient as similarity metric.
The Correlation does not care about the properties of the underlying timesieres.

- Do we still need to remove non-stationarity from the data:
Matrix Profile can operate on any time-series.
- Do matrix profile derived features make rolling window features (such as EMA) redundant:
No. Moving Average smoothes the input data. Matrix Profile searches for the nearest-neighbor-sequence of a query-sequence.
- What are the examples of using Matrix profiles in conjunction with supervised learning models:
If you have known patterns/motifs you can construct features for supervised learning.
Precompute the similarity vector with MASS(query-patter, time-series) for each known pattern.
The similarity vectors are your additional features.
- Can Matrix profiles be used as a standalone forecasting framework, without ML models. If so, how?
No. The result of matrix profile (the vector of distances) is not a direct forecast.

**Comment by seanlaw** [comment]
@dirtyValera Thank you for your question and welcome to the STUMPY community!

@JaKasb provided some useful responses but I wanted to provide some additional considerations:

Regarding non-stationarity, one thing to remember is that, by default in STUMPY, matrix profiles are computed by first z-normalizing each subsequence before comparing the pair-wise subsequence Euclidean distances. Essentially, this normalizes the amplitude of each subsequence being compared and results in a comparison of the shapes of the subsequences. So, you aren't required to, say, diff your time series first before computing the matrix profile but it is possible that doing so beforehand may be useful. You'll want to test-and-learn but I usually just run `stumpy.stump` on raw time series. If amplitude matters to you then you'll want to set `normalize=False` when calling relevant STUMPY functions and this computes the Euclidean distance without z-normalization.

While not exactly time series forecasting, in one of the original [matrix profile papers](https://www.cs.ucr.edu/~eamonn/chains_ICDM.pdf) they describe a technique called "Time Series Chains" that somewhat resemble forecasting. You can read more about it in our [Time Series Chains Tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Time_Series_Chains.html#Retrieve-the-Data). 

In our experience, the success of the approach largely depends on your data. However, after you compute the matrix profile, time series chains are very cheap to compute and so it's worth a quick test. I hope this helps.
